### ISSUE TITLE
Frescobaldi becomes very slow on large files: use background thread for music related parsing/computation while editing

### ISSUE THREAD

**Issue by wbsoft** [wbsoft/frescobaldi#473]
Heavy computations that are done often should move to a background thread.

The ly.music module is used after every document change to parse the whole (tokenized) file and provide information such as the length of a selected music fragment or the time position of the text cursor in the status bar. This makes the editor feel sluggish on larger LilyPond files.

While editing a document, QSyntaxHighlighter triggers rehighlighting of blocks that drive the tokenizer, which does not work in a background thread but is fast enough. So we always have the tokens up-to-date. But building a `ly.music` tree of the tokenized text should be done in a background thread, also preferable only for the top-level expression that was changed.

Then the music events are traversed to measure time etc. This may traverse a lot of nodes and even load included files (although they are cached). So also those computations should go to the background, so that the editor does not lag behind when a user is typing quickly.

**Comment by uliska** [comment]
I just started to read about threading in Python and learned that this is not always efficient because Python itself runs in a single thread. So for at least some of our time-consuming repeated operations it may not deliver as significant improvements as desired. (From a recent experience I can confirm that the slow response can be a showstopper for potential users and use cases.)

Would it be an option to start up a server in a separate process to handle that kind of analysis tasks?

**Comment by uliska** [comment]
With the new JobQueue code in place it might be an incentive to do parsing for both highlighting and autocompletion in a background *process* rather than a *thread*. It's `done` signal would then update a data structure which would simply be somewhat out-of-date before that.

Also, when the parsing is triggered by a change in the open document it should by default *not* follow the includes (except if the change adds a new include file).

**Comment by gilbertohasnofb** [comment]
I know this is an old issue but the problem seems to still be valid. I am currently editing a very large file (around 10k lines) and Frescobaldi lags for a couple of seconds whenever I type a `-` or a `\`, probably due to the autocompletion issues raised before. If I directly paste something such as `--` for a tenuto or `\shape` for tweaking slurs, Frescobaldi's response is snappy.

**Comment by ksnortum** [comment]
The lag problem seems to have been exacerbated by the PyQt6 port.

**Comment by bmjcode** [comment]
@ksnortum Does 5cf20d56f95ed03703e54c1ad666053490162f7b help any? I know the time position's not the main culprit but it was a relatively easy first step.

**Comment by ksnortum** [comment]
Yes, it seems to help a lot, just based on my perfunctory testing.

**Comment by ksnortum** [comment]
There is still lag sometimes when editing in version 4.0.1.  The problem seems to not only be files with many lines (~1500 and up) but also how long you have been editing.  If you shutdown Frescobaldi and restart it, the lag seems to go away for a while.  This seems more like a memory leak than a background processing problem -- but it could be both.

**Comment by bameylan** [comment]
I have exactly the same problem with a 64 kb .ly file, so not too large a file. Frescobaldi becomes slower and slower over time. Actually, LilyPond compiles normally, but it's Frescobaldi that slows down the whole system. Typing notes takes time and if I type too quickly, a number of notes simply don't appear. The system monitor of my Linux shows me for Frescobaldi: **98 Mb** of memory used at the start of the work and **15 Gb** after about an hour of work.

**Comment by ksnortum** [comment]
@bameylan can you copy the version text from Help -> About Frescobaldi -> Version tab and paste it here?

**Comment by bameylan** [comment]
Here:
Frescobaldi: 4.0.4
Extension API: 0.9.0
Python: 3.12.12
python-ly: 0.9.9
Qt: 6.9.3
PyQt: 6.9.1
qpageview: 1.0.2
OS: Linux-6.14.0-37-generic-x86_64-with-glibc2.40 (KDE Flatpak runtime)
Installation kind: Flatpak

**Comment by fedelibre** [comment]
If you need a +2000 lines public score to reproduce this bug, [check this Chopin ballade](https://github.com/MutopiaProject/MutopiaProject/pull/1139).

**Comment by bameylan** [comment]
@fedelibre I can't compile this .ly file with my version of LilyPond… lots of errors…

**Comment by fedelibre** [comment]
The version in that PR compiles fine with version 2.24 of LilyPond, which is available in the flatpak package you are using.

**Comment by ksnortum** [comment]
If it's any help, [this project on GitHub](https://github.com/ksnortum/chopin-opus-54) has a file with over 5000 lines.  It's also written for LilyPond 2.24.

**Comment by bameylan** [comment]
Ok, it's done and without any noticeable problems. As the file is quite large, it takes a little time, but nothing comparable to the problem I have with my version of Frescobaldi: it slows down over time, especially when entering notes, and it's not the compilation that's the problem.

**Comment by bmjcode** [comment]
~#2145 adds a toggle under Preferences -> Editor -> Show music position in the status bar. See if turning that off helps any. I think it should since updating the music position was identified before as a major source of lag.~

Scratch that, see below.

**Comment by bmjcode** [comment]
~I had another idea this morning for how to solve this. See #2147.~ Nope, that didn't pan out.

**Comment by bmjcode** [comment]
After another morning of messing around with this, I've gotten it to lag in a different spot than before:

https://github.com/bmjcode/frescobaldi/tree/delay-slow-operations

More seriously, I've learned from profiling that `DocumentInfo.music()` is the main bottleneck. Right now its data is regenerated the first time `music()` is called after any change to the document. My idea here was that the *change itself* should trigger regeneration so `music()` just has to return the cached results. Since it would be counterproductive to do this after every individual keystroke, I adapted the existing code from `MusicPosition` to determine when the user has stopped typing.

I had hoped clever timing here would avoid the need to introduce threading and all the complications it entails. Alas, it looks like threading will still be necessary to avoid locking up the interface. This may give us a better foundation to do it on though...

*(Edited to more thoroughly explain the logic here.)*

**Comment by bmjcode** [comment]
Further experiments reveal we have **two** bottlenecks:

`DocumentInfo.music()` is mostly used to calculate the music position, and can be processed in a background thread relatively easily.

`DocumentInfo.lydocinfo()` is used by autocomplete, among many other things. So far attempting to process it in a background thread has introduced all manner of interesting breakage, starting with the cursor jumping around erratically with every keystroke. It also does nothing to help with lag (in fact, it seems to cause even more).

That's about all I can coherently explain right now. The one thing I've learned for sure here is how little I really understand this code.

**Comment by bmjcode** [comment]
I've had a breakthrough. It turns out the highlighter is responsible for tokenization as well as highlighting text. When the document moves between threads, the highlighter needs to move with it. Doing that has fixed most of the breakage I observed last night.

This experiment is starting to show promise. I'm going to move my branch to the main repo so it's easier to test.
